framework/src/utils/OffDiagonalScalingMatrix.C

```diff
@@ -0,0 +1,107 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "OffDiagonalScalingMatrix.h"
+#include "MooseError.h"
+
+#include "libmesh/auto_ptr.h"
+#include "libmesh/numeric_vector.h"
+#include <cmath>
+
+using namespace libMesh;
+
+template <typename T>
+OffDiagonalScalingMatrix<T>::OffDiagonalScalingMatrix(const Parallel::Communicator & comm_in)
+  : DiagonalMatrix<T>(comm_in)
+{
+}
+
+template <typename T>
+std::unique_ptr<SparseMatrix<T>>
+OffDiagonalScalingMatrix<T>::zero_clone() const
+{
+  // Make empty copy with matching comm
+  auto mat_copy = libmesh_make_unique<OffDiagonalScalingMatrix<T>>(this->comm());
+
+  // Initialize copy with our same nonzero structure, and explicitly
+  // zero values using fast == false.
+  mat_copy->init(*this, /*fast=*/false);
+
+  // Work around an issue on older compilers.  We are able to simply
+  // "return mat_copy;" on newer compilers
+  return std::unique_ptr<SparseMatrix<T>>(mat_copy.release());
+}
+
+template <typename T>
+std::unique_ptr<SparseMatrix<T>>
+OffDiagonalScalingMatrix<T>::clone() const
+{
+  // Make empty copy with matching comm
+  auto mat_copy = libmesh_make_unique<OffDiagonalScalingMatrix<T>>(this->comm());
+
+  // Make copy of our diagonal
+  auto diag_copy = this->_diagonal->clone();
+
+  // Swap diag_copy with diagonal in mat_copy
+  *mat_copy = std::move(*diag_copy);
+
+  // Work around an issue on older compilers.  We are able to simply
+  // "return mat_copy;" on newer compilers
+  return std::unique_ptr<SparseMatrix<T>>(mat_copy.release());
+}
+
+template <typename T>
+void
+OffDiagonalScalingMatrix<T>::set(const numeric_index_type, const numeric_index_type, const T)
+{
+  mooseError("We should never call set for a scaling matrix. Set is the kind of thing that is done "
+             "for constraints, which are generally not on the order of magnitude of the physics, "
+             "which is what we are looking at when determining scaling.");
+}
+
+template <typename T>
+void
+OffDiagonalScalingMatrix<T>::add(const numeric_index_type i,
+                                 const numeric_index_type /*j*/,
+                                 const T value)
+{
+  this->_diagonal->add(i, std::abs(value));
+}
```

Comment:
Intuitively, I think we should take the maximum col value instead of summing them up

Reply:
So like ‘if value > current_value, then current_value = value’? Otherwise we will need more storage

Reply:
Your Idea was to keep storage minimal and keep things “diagonal” haha

Reply:
I was thinking loudly. It seems NOT easy to figure out the maximum absolute value without a full storage. 

For each row, values can be added in any order, depending on a FEM loop we use. It is even worse that each add might involve only a partial value for a given DoF pair.  So the "if value > current_value, then current_value = value" method won't work at all. 

We have to stay with as is ....

Reply:
You mean avoid full storage of the matrix?

Reply:
I m thinking about how to minimize other norms beyond L_infty of each column

Reply:
I can not  have `L_infty` of each column without a full  storage of the matrix.

Reply:
same for L2 then right?
When we you mean we add a partial value, it s because this matrix is doing a partial storage before accumulating to another matrix?

Reply:
We cache individual kernel contributions separately and then add each one of these individual cached values separately

Reply:
Ok I dont know how to do better than L1 norm then. 
I ve been looking at https://web.stanford.edu/~takapoui/preconditioning.pdf and https://arxiv.org/pdf/1704.02315.pdf for other methods (not exactly the same problem) but seems we need the matrix for all of these.